### PR TITLE
feat: Airdrop contract

### LINKED ISSUE #78: [Contract] Airdrop Contract
## 🎯 Objective

Create a contract that automates the distribution of **XLM** or marketplace tokens to eligible users based on certain conditions, such as previous purchases, participation, or loyalty.

---

### 🏗 Contract structure

```rust
airdrop-contract/src/

  lib.rs             // Configuration and exports
  eligibility.rs     // Define eligibility criteria
  distribution.rs    // Airdrop logic and calculations
  tracking.rs        // User participation tracking
```

---

#### 1. Airdrop Core  
- Roles: **Admin**, **User**  
- Admin can trigger airdrop events based on eligibility criteria  
- Automated distribution to eligible users  
- Eligibility based on factors like past purchases, account activity, or level of loyalty  

#### 2. Airdrop Distribution  
- Airdrop distribution occurs in **XLM** or marketplace token  
- Users are credited automatically based on predefined criteria  
- Max cap per user (to prevent abuse)

#### 3. Eligibility Criteria  
- Past purchase history  
- Account activity (number of logins, referrals)  
- Loyalty program level or subscription status  

#### 4. Airdrop Transparency  
- Public ledger of all airdrops (on-chain)  
- Track which users received tokens and how much  
- Set frequency and amount for future airdrops  

#### 5. Anti-Abuse & Security  
- Prevent bots or fake accounts from claiming airdrops  
- Limit airdrops to prevent spam  
- Secure claim process with unique validation for each user

---

### 🔍 Clarity

- Clear definition of eligibility and participation rules  
- Transparent tracking of distribution  
- Airdrop event management and admin controls  

---

### ⏳ ETA

- 2 Days

---

### 🔗 References

**[Stellar Official Guide](https://developers.stellar.org/#smart-contract-developers) and [Rust Book](https://doc.rust-lang.org/book/).**

---

### 📌 Additional Notes

- Can be integrated with subscription or referral systems  
- Potential for “loyalty airdrops” for active users  
- Trigger airdrops based on specific milestones (holidays, marketplace events)

---

### PR DESCRIPTION
**Pull request by FrankiePower**
# 🚀 StarShop Pull Request

### **Pull Request: Implement Airdrop Contract for Automated Token Distribution on Stellar**

---

### 🧠 Summary

This PR implements a powerful and flexible `AirdropContract` on Soroban that allows automated token distribution based on customizable eligibility conditions. It supports both XLM and custom tokens, enables admin-driven and user-initiated claims, and integrates easily with other systems like referrals or purchases via dynamic metrics.

---

### 📌 Type of Change

- [ ] Documentation (updates to README, docs, or comments)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

---

#### **Context**

This PR introduces the `AirdropContract`, a Stellar smart contract built with Soroban SDK to automate the distribution of XLM or marketplace tokens to eligible users based on customizable conditions such as past purchases, account activity, or loyalty. The contract supports admin-driven airdrop events, user claims, and batch distribution, with a focus on transparency, security, and scalability. It leverages Stellar’s native token handling (`token::Client`) and persistent storage for on-chain tracking, ensuring compatibility with the Stellar blockchain.

The contract is modular, comprising:

- `lib.rs`: Main interface for initialization, triggering airdrops, claiming, and batch distribution.

- `eligibility.rs`: Handles dynamic eligibility checks.

- `distribution.rs`: Manages token transfers.

- `tracking.rs`: Tracks user metrics.

- `types.rs`: Defines shared data structures with dynamic conditions and metrics.

- `external.rs`: Defines an interface that external contracts can implement for the airdrop contract to get metrics.

---

#### **Assumptions**

- **Token Handling**:
  - XLM and custom tokens are treated uniformly using `token::Client`, with the token address provided in `AirdropEvent`. The XLM token address is assumed to be the native asset’s address on Stellar.
  - Tokens must be transferred to the contract before distribution (e.g., via off-chain funding or marketplace integration).

- **User Data Source**:
  - User metrics (e.g., purchases, activity points) are updated by the admin via `external.rs`, assuming an external system (e.g., marketplace, referral system) provides accurate data.
  - No on-chain user registry exists; `distribute_all` requires a provided user list, which can be sourced off-chain or extended with on-chain storage.

- **Stellar Environment**:
  - The contract operates on Stellar’s Soroban SDK, using persistent storage for events and claims.
  - `token::Client` handles all token interactions (XLM and custom), with no special XLM methods (e.g., `.xlm`).

- **Test Environment**:
  - Tests will use Soroban’s test utilities (e.g., `soroban_sdk::testutils`) to simulate the Stellar ledger, token contracts, and user interactions.
  - Assumes standard behavior for `require_auth`, storage, and event emission.

- **External Integrations**:
  - The contract can integrate with subscription or referral systems by extending `UserData.metrics` with new keys (e.g., `"subscriptions"`, `"referrals"`).
  - Milestone-based airdrops (e.g., holidays) are supported via flexible `claim_airdrop`.

---

####  📝 Changes description

1. **Contract Initialization**:
   - `initialize` in `lib.rs` sets the admin address and event ID counter, ensuring single initialization.

2. **Airdrop Event Creation**:
   - `claim_airdrop` allows admins to create events with dynamic conditions (`Map<Symbol, u64>`), token address, and amount.
   - Supports XLM or custom tokens, with validation for positive amounts.

3. **Eligibility Checking**:
   - `eligibility.rs` checks user metrics against event conditions dynamically, supporting any criteria (e.g., purchases, loyalty).

4. **Token Distribution**:
   - `distribution.rs` handles token transfers using `token::Client`, with unified logic for XLM and custom tokens.
   - `claim_airdrop` allows user-initiated claims with eligibility checks and authentication.
   - `distribute_all` enables admin-driven batch distribution for multiple users.

5. **User Metric Tracking**:
   - `external.rs` supports fetching dynamic metrics.

6. **Transparency and Security**:
   - On-chain storage (`DataKey::AirdropEvent`, `DataKey::Claimed`) and events (`airdrop_claimed`, `claimed`) ensure a public ledger.
   - Anti-abuse measures include `require_auth`, claim limits, and admin-only controls.

---

#### **Design Considerations**

Our conversation shaped the contract’s design, addressing key challenges and aligning with the task’s requirements. Below are the major design considerations and why we implemented the contract this way:

1. **Dynamic Conditions and Metrics**:
   - **Challenge**: Support hundreds of admins with diverse conditions (e.g., purchases, referrals) without redeploying, given Stellar’s immutable contracts.
   - **Consideration**: Fixed fields (`min_purchases`, `loyalty_level`) or enums (`ConditionType`) limit flexibility and can’t be modified post-deployment.
   - **Chosen Approach**: Used `Map<Symbol, u64>` in `AirdropEvent.conditions` and `UserData.metrics` for dynamic conditions and metrics, allowing any criteria (e.g., `"purchases"`, `"subscriptions"`) without code changes.
   - **Why**: Ensures scalability and future-proofing, supporting the task’s requirement for flexible eligibility criteria and integration with referral/subscription systems. Tested in `test_check_eligibility`.

2. **Unified Token Handling**:
   - **Challenge**: Correctly handle XLM and custom tokens, avoiding incorrect assumptions (e.g., `.xlm` method).
   - **Consideration**: Initial `distribution.rs` used a non-existent `.xlm` method, inconsistent with Stellar’s token model.
   - **Chosen Approach**: Adopted `token::Client` for all transfers (XLM and custom tokens), with `AirdropEvent.token_address` specifying the token. Fixed based on `TransactionContract`’s unified approach.
   - **Why**: Aligns with Stellar’s design, simplifies logic, and ensures compatibility with XLM as a native token. Tested in `test_distribute_tokens`.

3. **Automated vs. User-Initiated Distribution**:
   - **Challenge**: Balance automation with gas efficiency and user control.
   - **Consideration**: Fully automated distribution (e.g., airdropping to all users on-chain) requires iterating over a user list, which can be gas-intensive.
   - **Chosen Approach**: Implemented both `claim_airdrop` (user-initiated) and `distribute_all` (admin-driven batch distribution), with `distribute_all` requiring a provided user list to avoid on-chain iteration.
   - **Why**: Provides automation via `distribute_all` while keeping `claim_airdrop` for user flexibility, meeting the task’s automation goal. Off-chain user lists reduce gas costs. Tested in `test_distribute_all`.

4. **Transparency and Public Ledger**:
   - **Challenge**: Ensure all airdrop activities are auditable on-chain.
   - **Consideration**: Stellar’s blockchain is public, but contract data must be structured for easy access.
   - **Chosen Approach**: Stored `AirdropEvent` and `Claimed` data in persistent storage (`DataKey`) and emitted detailed events (`airdrop_triggered`, `claimed`) with token address and amount.
   - **Why**: Meets the task’s transparency requirement, providing a verifiable ledger. Events enhance off-chain tracking. Tested in `test_transparency`.

5. **Anti-Abuse and Security**:
   - **Challenge**: Prevent bots, fake accounts, and spam claims.
   - **Consideration**: Unrestricted claims or metric updates could allow abuse.
   - **Chosen Approach**: Implemented `require_auth` in `claim_airdrop`, one-claim-per-event limits (`DataKey::Claimed`), admin-only metric updates (`tracking.rs`), and amount validation (`InvalidAmount`).
   - **Why**: Ensures secure claims and data integrity, aligning with the task’s anti-abuse requirements. Could extend with account age checks if needed. Tested in `test_security`.

6. **Scalability for Conditions**:
   - **Challenge**: Support a “menu” of conditions for admins, scalable to hundreds of criteria.
   - **Consideration**: Enums (`ConditionType`) are immutable post-deployment, limiting new conditions.
   - **Chosen Approach**: Used `Map<Symbol, u64>` to allow any condition name, avoiding enums and enabling new criteria without redeployment.
   - **Why**: Provides the flexibility requested for diverse admin needs and milestone-based airdrops. Tested in `test_trigger_airdrop`.

7. **Clarity and Modularity**:
   - **Challenge**: Maintain a clear, maintainable codebase.
   - **Consideration**: Complex logic could obscure eligibility or distribution rules.
   - **Chosen Approach**: Split functionality into modules (`eligibility.rs`, `distribution.rs`, etc.), with clear interfaces and dynamic data structures.
   - **Why**: Enhances readability and maintainability, meeting the task’s clarity requirement. Tested across all modules.

---

#### **Acceptance Criteria**

1. **Airdrop Core**:
   - ✅ Admin triggers events via `create_airdrop_event`.
   - ✅ Users claim via `claim_airdrop`.
   - ✅ Automated distribution via `distribute_all`.
   - ✅ Eligibility supports purchases, activity, loyalty, and more.

2. **Airdrop Distribution**:
   - ✅ Supports XLM and custom tokens via `token::Client`.
   - ✅ Automatic crediting in `claim_airdrop` and `distribute_all`.
   - ✅ Max cap enforced by `AirdropEvent.amount`.

3. **Eligibility Criteria**:
   - ✅ Dynamic `Map<Symbol, u64>` supports all specified criteria and extensibility.

4. **Airdrop Transparency**:
   - ✅ On-chain storage and events create a public ledger.
   - ✅ Tracks recipients and amounts via `DataKey::Claimed` and `claimed` events.
   - ✅ Admins can set amounts for future airdrops.

5. **Anti-Abuse & Security**:
   - ✅ Prevents bots with `require_auth` and admin-only metric updates.
   - ✅ Limits claims to one per event.
   - ✅ Secure claims with authentication and eligibility checks.

6. **Clarity**:
   - ✅ Clear eligibility rules in `eligibility.rs`.
   - ✅ Transparent tracking via events and storage.
   - ✅ Admin controls via `trigger_airdrop`, `distribute_all`.

7. **Additional Notes**:
   - ✅ Supports integration with subscriptions/referrals via dynamic metrics.
   - ✅ Enables loyalty airdrops with flexible conditions.
   - ✅ Triggers milestone airdrops via `trigger_airdrop`.

---

#### 📸  **Evidence **

**On `cargo build`**:
Screenshot from 2025-04-30 14-15-45

**On `cargo test`**:
Screenshot from 2025-04-30 14-15-13

---

## ⏰ Time spent breakdown

**Total**: 3 full days (~24–30 focused hours)

| Day     | Focus Area         | Tasks |
|---------|--------------------|-------|
| **Day 1** | 🧠 Design & Architecture | Scoped out feature set, designed modular structure, defined data types and error enums, outlined external contract interfaces. |
| **Day 2** | 🔨 Core Implementation | Implemented `lib.rs`, distribution logic, eligibility checks, tracking, and event lifecycle flows. Integrated token transfers and metrics providers. |
| **Day 3** | ✅ Testing & Hardening | Wrote unit tests, verified behavior manually with Soroban CLI, tested edge cases, reviewed gas efficiency, finalized error handling and event emissions. |

---

#### **Checklist**

- [x] Implement airdrop event creation and claiming.
- [x] Add dynamic eligibility checking with extensible criteria.
- [x] Implement token distribution for XLM and custom tokens.
- [x] Support user claims and batch distribution.
- [x] Ensure transparency with on-chain storage and events.
- [x] Add anti-abuse measures (authentication, claim limits).
- [x] Write modular, clear code per task structure.
- [x] Write unit tests for all functions.
- [x] Verify functionality with tests.
- [x] Run tests
- [x] Run formatting
- [x] Evidence attached
- [x] Commented the code
- [x] Closes #78

### CODE REVIEW

File: StarShopContracts/airdrop-contract/src/test.rs [hunk 1]
```diff
@@ -0,0 +1,1314 @@
...
+use referral_contract::{ReferralContract, types::UserData};
+
+fn create_test_env() -> (Env, Address) {
+    let env = Env::default();
+    env.mock_all_auths();
+    let airdrop_contract = env.register(AirdropContract, ());
+    (env, airdrop_contract)
+}
+
+pub fn setup_contract(e: &Env) -> (ReferralContractClient, Address, Address, Address) {
```
Comment (aguilar1x) [inline StarShopContracts/airdrop-contract/src/test.rs:26]: change parameter e, by env pls

**Review by aguilar1x — CHANGES_REQUESTED** [review]
good implementation, but don't change the referral contract, so revert those changes

File: StarShopContracts/airdrop-contract/src/test.rs (hunk 1 above)
Comment (FrankiePower) [reply StarShopContracts/airdrop-contract/src/test.rs:26]: yeah i know that. i am still working on the tests.

**Comment by FrankiePower** [comment]
@aguilar1x PR is ready to be reviewed.

File: StarShopContracts/referral-contract/Cargo.toml [hunk 2]
```diff
@@ -5,7 +5,6 @@ edition = "2021"
 publish = false
 
 [lib]
-crate-type = ["cdylib"]
```
Comment (aguilar1x) [inline StarShopContracts/referral-contract/Cargo.toml:8]: Don't delete this

File: Cargo.toml [hunk 3]
```diff
@@ -1,6 +1,6 @@
 [workspace]
 resolver = "2"
-members = [
+members = [ "StarShopContracts/airdrop-contract",
```
Comment (aguilar1x) [inline Cargo.toml:3]: add this one line below

File: StarShopContracts/referral-contract/Cargo.toml (hunk 2 above)
Comment (FrankiePower) [reply StarShopContracts/referral-contract/Cargo.toml:8]: yeah i did that to import the referral contract. i will revert it back

File: Cargo.toml (hunk 3 above)
Comment (FrankiePower) [reply Cargo.toml:3]: done

**Comment by FrankiePower** [comment]
@aguilar1x  All Done!

**Comment by FrankiePower** [comment]
i'm done!

File: StarShopContracts/airdrop-contract/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,16 @@
+[package]
+name = "airdrop-contract"
+version = "0.1.0"
+edition = "2024"
```
Comment (aguilar1x) [inline StarShopContracts/airdrop-contract/Cargo.toml:4]: change the versión at 2021 pls

Comment (FrankiePower) [reply StarShopContracts/airdrop-contract/Cargo.toml]: done

**Comment by FrankiePower** [comment]
@aguilar1x anything else chief?

**Review by aguilar1x — APPROVED** [review]
LGTM!
